src/org/zaproxy/zap/extension/ascanrulesAlpha/HttpOnlySite.java

```diff
@@ -156,58 +152,70 @@ public void scan() {
-				String loc = newRequest.getResponseHeader().LOCATION;
-				System.out.println("\n\nThe location is: "+newRequest.getResponseHeader().getHeader(loc));
 				int status = newRequest.getResponseHeader().getStatusCode();
-				String redirect = newRequest.getResponseHeader().getHeader(loc);
-				if(status == HttpStatusCode.MOVED_PERMANENTLY || status == HttpStatusCode.FOUND) {
-					String path = newRequest.getRequestHeader().getURI().getPath();
-					if(redirect.contains("https")) {
-						String newHost = redirect.substring(8, redirect.length());
-					    newRequest.getRequestHeader().setURI(new URI("https", null, newHost, 443, path));
-					    record = "https";
-					    count++;
-					} else if (redirect.contains("http")) {
-						raiseAlert(newRequest);
-						break;
+				if(HttpStatusCode.isRedirection(status)) {
+					String loc = HttpResponseHeader.LOCATION;
+					String redirect = newRequest.getResponseHeader().getHeader(loc);
+					if(!redirect.isEmpty() && redirect != null) {
+						URI oldURI = newRequest.getRequestHeader().getURI();
+						try {
+							newRequest.getRequestHeader().setURI(new URI(oldURI, redirect, true));
+							URI newURI = newRequest.getRequestHeader().getURI();
+							if(!oldURI.getHost().equals(newURI.getHost())) {
+								String message = "Different hosts";
+								raiseAlert(newRequest, message);
+								return;
+							}
+						} catch (URIException e) {
+							try {
+								newRequest.getRequestHeader().setURI(new URI(oldURI, redirect, false));
+							} catch (URIException ex) {
+								String message = "URL could not be encoded";
+								raiseAlert(newRequest, message);
+								return;
+							}
+						}
+						if(newRequest.getRequestHeader().getURI().getScheme().equals("https")) {
+						    scheme = "https";
+						    count++;
+						} else if (redirect.startsWith("http")) {
+							String message = "Redirection was to an HTTP page";
+							raiseAlert(newRequest, message);
+							return;
+						}	
 					}
-				}
-				else {
+				}else {
 					break;
 				}
 			}
-			if(record.equals("http") && count == 10) { //For cases when even the 10th redirection is not https
-				raiseAlert(newRequest);
+			if(scheme.equals("http") && count == 10) { //For cases when even the 10th redirection is not https
+				String message = "Redirection was to an HTTP page";
+				raiseAlert(newRequest, message);
+				return;
 			}
-		} catch(NoRouteToHostException e) {
-			System.out.println("NoRoute");
-			raiseAlert(newRequest);
-			return;
 		} catch (SocketException | SocketTimeoutException e) {
-			System.out.println("SocketException");
-			raiseAlert(newRequest);
+			String message = "Failed to connect";
+			raiseAlert(newRequest, message);
 			return; 
 		} catch (SSLException e) {
-			System.out.println("SSL");
 			if(e.getMessage().contains("plaintext")) {
-				raiseAlert(newRequest);
+				String message = "Site has no SSL/TSL certificate";
```

Comment:
There's a typo "TSL".
Maybe "connection" (or "support"?) instead of "certificate"?